contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenDestination.sol

```diff
@@ -0,0 +1,269 @@
+                IERC20(feeInfo.feeTokenAddress),
+                address(teleporterMessenger),
+                adjustedFeeAmount
+            );
+        }
+
+        // Burn native token by sending to BURN_FOR_TRANSFER_ADDRESS
+        payable(BURN_FOR_TRANSFER_ADDRESS).transfer(msg.value);
+
+        uint256 messageID = teleporterMessenger.sendCrossChainMessage(
+            TeleporterMessageInput({
+                destinationBlockchainID: sourceBlockchainID,
+                destinationAddress: nativeTokenSourceAddress,
+                feeInfo: feeInfo,
+                requiredGasLimit: TRANSFER_NATIVE_TOKENS_REQUIRED_GAS,
+                allowedRelayerAddresses: allowedRelayerAddresses,
+                message: abi.encode(
+                    ITokenSource.SourceAction.Unlock,
+                    abi.encode(recipient, msg.value)
+                )
+            })
+        );
+
+        emit TransferToSource({
+            sender: msg.sender,
+            recipient: recipient,
+            amount: msg.value,
+            teleporterMessageID: messageID
+        });
+    }
+
+    /**
+     * @dev See {INativeTokenDestination-reportTotalBurnedTxFees}.
+     */
+    function reportTotalBurnedTxFees(
+        TeleporterFeeInfo calldata feeInfo,
+        address[] calldata allowedRelayerAddresses
+    ) external {
+        uint256 totalBurnedTxFees = address(BURNED_TX_FEES_ADDRESS).balance;
+        uint256 messageID = teleporterMessenger.sendCrossChainMessage(
+            TeleporterMessageInput({
+                destinationBlockchainID: sourceBlockchainID,
+                destinationAddress: nativeTokenSourceAddress,
+                feeInfo: feeInfo,
+                requiredGasLimit: REPORT_BURNED_TOKENS_REQUIRED_GAS,
+                allowedRelayerAddresses: allowedRelayerAddresses,
+                message: abi.encode(
+                    ITokenSource.SourceAction.Burn,
+                    abi.encode(totalBurnedTxFees)
+                )
+            })
+        );
+
+        emit ReportTotalBurnedTxFees({
+            burnAddressBalance: totalBurnedTxFees,
+            teleporterMessageID: messageID
+        });
+    }
+
+    /**
+     * @dev See {INativeTokenDestination-isCollateralized}.
+     */
+    function isCollateralized() external view returns (bool) {
+        return currentReserveImbalance == 0;
+    }
+
+    /**
+     * @dev See {INativeTokenDestination-totalSupply}.
+     */
+    function totalSupply() external view returns (uint256) {
+        uint256 burned = address(BURNED_TX_FEES_ADDRESS).balance +
+            address(BURN_FOR_TRANSFER_ADDRESS).balance;
+        uint256 created = totalMinted + initialReserveImbalance;
+
+        // This scenario should never happen, but this check will prevent an underflow
+        // where the contract would return a garbage value.
+        require(
+            burned <= created,
+            "NativeTokenDestination: FATAL - contract has tokens unaccounted for"
+        );
```

Comment:
Calling out that under and overflow checks are actually handled by Solidity itself as of 0.8.0, so `created - burned` would revert if underflow ever occured either way.

I'd super minor, but I'd be in favor of removing this check to reduce the gas cost for callers of `totalSupply`, and also to reduce the cognitive load of reasoning about the contract in the future. When I was initially reviewing I thought it may not be a proper check because tokens can be burned prior to from the initial allocation prior to any being minted, and had to double check to convince myself it is correct.